### PR TITLE
Weighted loss for Joint Intent and Slot models

### PR DESCRIPTION
**Pull request by purn3ndu**
One can use `--class_balancing=weighted_loss` to turn on weighted loss.
This improves the accuracy of the models for both intent and slots.

### CODE REVIEW

File: collections/nemo_nlp/nemo_nlp/data/datasets/utils.py [hunk 1]
```diff
@@ -9,6 +9,7 @@
 import shutil
 import subprocess
 import sys
+import itertools
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/data/datasets/utils.py]: arrange the built-in imports alphabetically

File: collections/nemo_nlp/nemo_nlp/data/datasets/utils.py [hunk 2]
```diff
@@ -1021,6 +1024,66 @@ def __init__(self,
...
+
+        for mode in ['train', 'test']:
+
+            slot_file = f'{self.data_dir}/{mode}_slots.tsv'
+            with open(slot_file, 'r') as f:
+                slot_lines = f.readlines()
+
+            input_file = f'{self.data_dir}/{mode}.tsv'
+            with open(input_file, 'r') as f:
+                input_lines = f.readlines()[1:]
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/data/datasets/utils.py]: future work: we might want to have an option to specify how many first lines of the file to skip instead of just picking 1

File: collections/nemo_nlp/nemo_nlp/data/datasets/utils.py [hunk 3]
```diff
@@ -1021,6 +1024,66 @@ def __init__(self,
...
+            raw_slots, queries, raw_intents = [], [], []
+            for slot_line, input_line in dataset:
+                slot_list = [int(slot) for slot in slot_line.strip().split()]
+                raw_slots.append(slot_list)
+                parts = input_line.strip().split()
+                raw_intents.append(int(parts[-1]))
+                queries.append(' '.join(parts[:-1]))
+
+            print(raw_slots[:50])
+            print(raw_intents[:50])
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/data/datasets/utils.py]: is it for testing?

File: collections/nemo_nlp/nemo_nlp/data/datasets/utils.py [hunk 4]
```diff
@@ -1021,6 +1024,66 @@ def __init__(self,
...
+
+            infold = input_file[:input_file.rfind('/')]
+
+            logger.info(f'Three most popular intents during {mode}ing')
+            total_intents, intent_label_freq = get_label_stats(
+                raw_intents, infold + f'/{mode}_intent_stats.tsv')
+            merged_slots = itertools.chain.from_iterable(raw_slots)
+
+            logger.info(f'Three most popular slots during {mode}ing')
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/data/datasets/utils.py]: i lol-ed at {}-ing

File: collections/nemo_nlp/nemo_nlp/data/datasets/utils.py [hunk 5]
```diff
@@ -1021,6 +1024,66 @@ def __init__(self,
...
+                most_common_intent = intent_label_freq[0]
+                weighted_intents = sorted([(j, most_common_intent[1]/i)
+                                          for (j, i) in intent_label_freq])
+                self.intent_weights = [i for (_, i) in weighted_intents]
+                print(f'Intent weights are - {self.intent_weights}')
+
+            print('total_intents, intent_label_freq')
+            print(total_intents, intent_label_freq)
+            print('slots_total, slots_label_freq')
+            print(slots_total, slots_label_freq)
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/data/datasets/utils.py]: you might want to print them differently, e.g.
print('total_intents', total_intents)

also, you might want to looking into using logger

File: collections/nemo_nlp/nemo_nlp/modules/losses.py [hunk 6]
```diff
@@ -184,19 +185,37 @@ def create_ports():
...
+                 **kwargs):
         LossNM.__init__(self, **kwargs)
         self.num_slots = num_slots
-        self._criterion = nn.CrossEntropyLoss()
+        self.intent_loss_weight = intent_loss_weight
+
+        # For weighted loss to tackle class imbalance
+        if len(slot_classes_loss_weights) > 0:
+            self.slot_classes_loss_weights = \
+                torch.FloatTensor(slot_classes_loss_weights).cuda()
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/modules/losses.py]: 1. do we have to make sure that weights have the same number of elements as the number of slots?
2. don't just move it to cuda bc it won't be compatible on cpus. use self._device and move it to self._device

File: collections/nemo_nlp/nemo_nlp/modules/losses.py [hunk 7]
```diff
@@ -184,19 +185,37 @@ def create_ports():
...
+
+        if len(intent_classes_loss_weights) > 0:
+            self.intent_classes_loss_weights = \
+                torch.FloatTensor(intent_classes_loss_weights).cuda()
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/modules/losses.py]: same comments as slot_weights

File: collections/nemo_nlp/nemo_nlp/data/datasets/utils.py [hunk 8]
```diff
@@ -1021,6 +1024,66 @@ def __init__(self,
...
+            slots_total, slots_label_freq = get_label_stats(
+                merged_slots, infold + f'/{mode}_slot_stats.tsv')
+
+            if mode == 'train':
+
+                most_common_slot = slots_label_freq[0]
+                weighted_slots = sorted([(j, most_common_slot[1]/i)
+                                        for (j, i) in slots_label_freq])
+                self.slot_weights = [i for (_, i) in weighted_slots]
+                print(f'Slot weights are - {self.slot_weights}')
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/data/datasets/utils.py]: 1. it'd be helpful to have comments explaining how weights are calculated
2. they should be helper functions (atomic functions ftw)

File: examples/nlp/joint_intent_slot_with_bert.py [hunk 9]
```diff
@@ -131,6 +145,8 @@ def create_pipeline(num_samples=-1,
 
     intent_logits, slot_logits = classifier(hidden_states=hidden_states)
 
+    class_weights = data_desc.slot_weights
```
Comment (chiphuyen) [inline examples/nlp/joint_intent_slot_with_bert.py]: where is class_weights used?

File: collections/nemo_nlp/nemo_nlp/modules/losses.py (hunk 6 above)
Comment (purn3ndu) [reply collections/nemo_nlp/nemo_nlp/modules/losses.py]: 1. This is done automatically with the online calculation of weights unless the user makes some modifications manually.
2. Done. Let me know it its okay or should do it differently.

File: collections/nemo_nlp/nemo_nlp/data/datasets/utils.py [hunk 10]
```diff
@@ -1,21 +1,21 @@
-from collections import Counter
-from io import open
 import csv
 import glob
+import itertools
 import json
+import numpy as np
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/data/datasets/utils.py]: see style guide for imports
https://www.python.org/dev/peps/pep-0008/#imports

File: collections/nemo_nlp/nemo_nlp/data/datasets/utils.py [hunk 11]
```diff
@@ -936,6 +938,19 @@ def write_files(data, outfile):
...
 
 
+def calc_class_weights(label_freq):
+    # Goal is to give more weight to the classes with less samples
+    # so as to match the one with the higest frequency. We achieve this by
+    # dividing the highest frequency by the freq of each label.
+    # Example -
+    # [12, 5, 3] -> [12/12, 12/5, 12/3] -> [1, 2.4, 4]
+
+    most_common_label_freq = label_freq[0]
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/data/datasets/utils.py]: either:
1. check that label_freq is sorted
2. update docstring to show that label_freq has to be sorted so that label_freq[0] is the most frequent element

also, use """ for docstring

File: collections/nemo_nlp/nemo_nlp/modules/losses.py [hunk 12]
```diff
@@ -184,19 +185,37 @@ def create_ports():
...
                        slot_logits,
                        loss_mask,
                        intents,
-                       slots,
-                       intent_loss_weight=0.6):
-        intent_loss = self._criterion(intent_logits, intents)
+                       slots):
+        intent_loss = self._criterion_intent(intent_logits, intents)
 
         active_loss = loss_mask.view(-1)
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/modules/losses.py]: loss_mask should be int or long now to conform to the new version of pytorch, so you active_loss needs to be calculated the following way:
`active_loss = loss_mask.view(-1) > 0.5`
